src/Lean/Elab/Tactic/BVAckermannize.lean

```diff
@@ -0,0 +1,447 @@
+expressions and ackermannizes them in turn.
+-/
+
+partial def introAckForExpr (g : MVarId) (e : Expr) : AckM (AckResult × MVarId) :=
+  withContext g do
+    withTraceNode g (m!"🎯 {e}") (collapsed := false) do
+    match e with
+    | .mdata _ e => introAckForExpr g e
+    | .bvar _ => return ({ val := e, hasLooseBvar := true }, g)
+    | .proj tyName struct e =>
+      let (out, g) ← introAckForExpr g e
+      return ({ val := .proj tyName struct out.val, hasLooseBvar := out.hasLooseBvar}, g)
+    | .fvar .. | .mvar .. | .sort .. | .const .. |  .lit .. => do
+      return ({ val := e, hasLooseBvar := false }, g)
+    | .lam _binderName binderTy body _binderInfo =>
+        let (binderTy, g) ← introAckForExpr g binderTy
+        let (body, g) ← introAckForExpr g body
+        let val := e.updateLambdaE! binderTy.val body.val
+        let hasLooseBvar := binderTy.hasLooseBvar || body.hasLooseBvar
+        return ({ val, hasLooseBvar}, g)
+    | .letE _declName type value body _nonDep =>
+        let (type, g) ← introAckForExpr g type
+        let (value, g) ← introAckForExpr g value
+        let (body, g) ← introAckForExpr g body
+        let val := e.updateLet! type.val value.val body.val
+        let hasLooseBvar := type.hasLooseBvar || value.hasLooseBvar || body.hasLooseBvar
+        return ({ val, hasLooseBvar}, g)
+    | .forallE _binderName binderTy body _binderInfo =>
+        let (binderTy, g) ← introAckForExpr g binderTy
+        let (body, g) ← introAckForExpr g body
+        let val := e.updateForallE! binderTy.val body.val
+        let hasLooseBvar := binderTy.hasLooseBvar || body.hasLooseBvar
+        return ({ val, hasLooseBvar}, g)
+    | .app .. => withTraceNode g m!"🎯 Expr.app '{e}'" (collapsed := false) do
+      ackApp g e
+end
+
+/--
+Return true if the argument lists are trivially different.
+This is an optimization that we do not yet implement.
+-/
+def areArgListsTriviallyDifferent (_arg₁ _arg₂ : Array Argument) : AckM Bool := return false
+
+/-
+Return true if the argument lists are trivially the same.
+This is an optimization that we do not yet implement.
+If possible, return the simplified hypothesis of the equality of these expressions.
+TODO: -- def areArgListsTriviallySame (arg₁ arg₂ : Array Argument) : AckM (Option Expr) := return none
+-/
+
+
+/-- info: congr.{u, v} {α : Sort u} {β : Sort v} {f₁ f₂ : α → β} {a₁ a₂ : α} (h₁ : f₁ = f₂) (h₂ : a₁ = a₂) : f₁ a₁ = f₂ a₂ -/
+#guard_msgs in #check congr
+
+/-- This is how we build the congruence lemma for `n` arguments. -/
+private example (f : X1 → X2 → O)
+    -- we have these already.
+    (x1 x1' : X1) (x2 x2' : X2)
+    (ackEqApp : fx1x2 = f x1 x2)
+    (ackEqApp' : fx1'x2' = f x1' x2')
+    -- to be intros'd
+    (h1 : x1 = x1')
+    (h2 : x2 = x2') :
+  fx1x2 = fx1'x2' :=
+  let appEqApp : f x1 x2 = f x1' x2' := congr (congr (Eq.refl f) h1) h2
+  Eq.trans (Eq.trans ackEqApp appEqApp) (Eq.symm ackEqApp')
```

Comment:
This seems like they were notes for the implementation, but do we want to keep them around?